dex/farm/src/lib.rs

```diff
@@ -229,11 +248,63 @@ pub trait Farm:
             &farm_attributes,
         );
         Ok(MultiResult2::from((
-            self.create_payment(&farming_token_id, 0, &initial_farming_token_amount),
+            self.create_payment(&farm_token_id, nft_nonce, &initial_farming_token_amount),
             self.create_payment(&reward_token_id, reward_nonce, &reward),
         )))
     }
 
+    #[payable("*")]
+    #[endpoint(unbondFarm)]
+    fn unbond_farm(
+        &self,
+        #[payment_token] payment_token_id: TokenIdentifier,
+        #[payment_nonce] token_nonce: Nonce,
+        #[payment_amount] amount: BigUint,
+        #[var_args] opt_accept_funds_func: OptionalArg<ManagedBuffer>,
+    ) -> SCResult<()> {
+        require!(self.is_active(), "Not active");
+        require!(!self.farm_token_id().is_empty(), "No farm token");
+
+        let farm_token_id = self.farm_token_id().get();
+        require!(payment_token_id == farm_token_id, "Bad input token");
+        require!(amount > 0, "Payment amount cannot be zero");
+
+        self.farming_token_total_liquidity()
+            .update(|liq| *liq -= &amount);
+
+        let token_info = self.blockchain().get_esdt_token_data(
+            &self.blockchain().get_sc_address(),
+            &farm_token_id,
+            token_nonce,
+        );
+        let full_unlock_epoch = token_info
+            .decode_attributes::<UnbondSftAttributes>()?
+            .full_unlock_epoch;
+        let current_epoch = self.blockchain().get_block_epoch();
+
+        let farming_token_id = self.farming_token_id().get();
+        let farming_token_amount = if current_epoch < full_unlock_epoch {
```

Comment:
let's remove penalty completely. this is not needed in staking contract

Reply:
Done. Now the unbond period is forced, with no early unbond allowed.